[Identity] Support for tenant Id Challenges / tenant discovery in ClientCredentials (#15837)
Azure/azure-sdk-for-js: sdk/identity/identity/CHANGELOG.md

  - Added `regionalAuthority` property to `ClientSecretCredentialOptions` and `ClientCertificateCredentialOptions`.
  - If instead of a region, `AutoDiscoverRegion` is specified as the value for `regionalAuthority`, MSAL will be used to attempt to discover the region.
  - A region can also be specified through the `AZURE_REGIONAL_AUTHORITY_NAME` environment variable.
- The `AzureCliCredential` now receives an `AzureCliCredentialOptions` object, which has the same structure of the `TokenCredentialOptions`, but with an extra optional `tenantId`.
- The `AzurePowerShellCredential` now receives an `AzurePowerShellCredentialOptions` object, which has the same structure of the `TokenCredentialOptions`, but with an extra optional `tenantId`.
- All of the credentials now accept `allowMultiTenantAuthentication`, and as long as they allow configuring a `tenantId`, they now also work with the `GetTokenOptions` new property: `tenantId`.

>   - If a tenant is specified, any `getToken` request that receives a `tenantId` that doesn't match the credential's `tenantId` will throw an error, unless `getToken` gets called with the `allowMultiTenantAuthentication` property set to true.

Reviewer: ```suggestion
  - If a tenant is specified, any `getToken` request that receives a `tenantId` that doesn't match the credential's `tenantId` will throw an error, unless the credential has been constructed with the `allowMultiTenantAuthentication` option set to true.
```

  - Added `regionalAuthority` property to `ClientSecretCredentialOptions` and `ClientCertificateCredentialOptions`.
  - If instead of a region, `AutoDiscoverRegion` is specified as the value for `regionalAuthority`, MSAL will be used to attempt to discover the region.
  - A region can also be specified through the `AZURE_REGIONAL_AUTHORITY_NAME` environment variable.
- The `AzureCliCredential` now receives an `AzureCliCredentialOptions` object, which has the same structure of the `TokenCredentialOptions`, but with an extra optional `tenantId`.
- The `AzurePowerShellCredential` now receives an `AzurePowerShellCredentialOptions` object, which has the same structure of the `TokenCredentialOptions`, but with an extra optional `tenantId`.

> - All of the credentials now accept `allowMultiTenantAuthentication`, and as long as they allow configuring a `tenantId`, they now also work with the `GetTokenOptions` new property: `tenantId`.

Author: There’s a special case for `ManagedIdentityCredential`, which may support multi tenant authentication in the future. It currently doesn’t allow specifying the `tenantId`. We intentionally don’t want to bring attention to the `ManagedIdentityCredential` until we have a definite approach there.

> - All of the credentials now accept `allowMultiTenantAuthentication`, and as long as they allow configuring a `tenantId`, they now also work with the `GetTokenOptions` new property: `tenantId`.

Author: It should work on all credentials, but on ManagedIdentityCredential it doesn’t work yet (there’s work happening in the background to get us to have a good story for that credential, which won’t be done soon enough to wait). We’ll have a meeting soon to see if we should remote the property from ManagedIdentityCredential. In the mean time, I’ll keep the changelog weird.

  - Added `regionalAuthority` property to `ClientSecretCredentialOptions` and `ClientCertificateCredentialOptions`.
  - If instead of a region, `AutoDiscoverRegion` is specified as the value for `regionalAuthority`, MSAL will be used to attempt to discover the region.
  - A region can also be specified through the `AZURE_REGIONAL_AUTHORITY_NAME` environment variable.
- The `AzureCliCredential` now receives an `AzureCliCredentialOptions` object, which has the same structure of the `TokenCredentialOptions`, but with an extra optional `tenantId`.
- The `AzurePowerShellCredential` now receives an `AzurePowerShellCredentialOptions` object, which has the same structure of the `TokenCredentialOptions`, but with an extra optional `tenantId`.

> - All of the credentials now accept `allowMultiTenantAuthentication`, and as long as they allow configuring a `tenantId`, they now also work with the `GetTokenOptions` new property: `tenantId`.

Reviewer 2: I kinda want to merge all three of these bullets together, since it is a bit hard to piece together what is supported right now

Author: Hmm I’m not sure if I can see clearly how to merge these thee bullets.

Reviewer 2: I don't know, something like "`AzureCliCredential` and `AzurePowerShellCredential` now take an optional `tenantId` property, which enables the use of a new `allowMultiTenantAuthentication` option."

Author: well, the issue is that all the credentials do now! With one exception: ManagedIdentityCredential doesn’t take a `tenantId` (yet), this is because we haven’t come up with the solution for that credential across languages

> - The `AzureCliCredential` now receives an `AzureCliCredentialOptions` object, which has the same structure of the `TokenCredentialOptions`, but with an extra optional `tenantId`.
> - The `AzurePowerShellCredential` now receives an `AzurePowerShellCredentialOptions` object, which has the same structure of the `TokenCredentialOptions`, but with an extra optional `tenantId`.
> - All of the credentials now accept `allowMultiTenantAuthentication`, and as long as they allow configuring a `tenantId`, they now also work with the `GetTokenOptions` new property: `tenantId`.
>   - If a tenant is specified, any `getToken` request that receives a `tenantId` that doesn't match the credential's `tenantId` will throw an error, unless the credential has been constructed with the `allowMultiTenantAuthentication` option set to true.
>   - If multi-tenant authentication is allowed, a `tenantId` received through `GetTokenOptions` will be used for that specific request for token instead of the `tenantId` configured when the credential was initialized.

Reviewer 3: How about:
```suggestion
- `AzureCliCredential` and `AzurePowerShellCredential` now allow specifying a `tenantId`.
- All credentials except `ManagedIdentityCredential` support enabling multi tenant authentication via the `allowMultiTenantAuthentication` option.
```

Is that accurate? I think getting into the minutiae of explaining the relationship between the tenantId in credential options versus the tenantId in getToken options within the changelog is what's leading to confusion. I think that can be left explained in the docs.

Author: I love it! I’ll take it!
